reps/2024-05-21-ray-kubectl-plugin.md

```diff
@@ -0,0 +1,95 @@
+
+For novice Kubernetes users, an intuitive CLI experience can drastically enhance their onboarding journey. A kubectl plugin is proposed since it can seamlessly integrate with the rest of the existing kubectl surface as needed.
+The primary goal of this plugin should be user-friendliness and a smooth progression from zero to “something good enough”. For more advanced scenarios and scaling requirements, users can opt to manage KubeRay custom resources
+independently as many users do today.
+
+More simply, the Ray Kubectl Plugin simplifies the management of Ray clusters by eliminating the need for users to handle complex YAML configurations.
+
+Why a kubectl plugin instead of the existing kuberay CLI?
+* **Convenience / ease of use**: Kubectl plugins are designed to work within the existing kubectl framework. They inherit the user's current authentication, context, and cluster metadata, eliminating the need for additional setup.
+* **Seamless kubectl integration**: Users can effortlessly switch between basic kubectl subcommands and the Ray plugin using a single command-line tool.
+* **Accessibility**: The Kuberay CLI needs the KubeRay API server to function. However, the majority of clusters using KubeRay don't run the KubeRay API server, which limits the accessibility of the CLI.
+
+## User Journey
+
+### Create / Manage a Ray cluster
+
+```
+$ kubectl ray cluster create my-cluster --ray-version=2.9.3
+    --worker-replicas=10 --worker-cpus=8 -worker-memory=32GB --worker-gpus=2
+```
+
+```
+kubectl ray cluster scale default-worker-group --cluster my-cluster --replicas=10
+```
+
+```
+kubectl ray cluster worker-group create cpu-pool --cluster my-cluster -worker-replicas=10
+    --worker-cpus=8 -worker-memory=32GB
+
+```
+
+```
+kubectl ray cluster delete my-cluster
+```
+
+### Ray Job Submissions (using RayJob)
+
+```
+$ kubectl ray job submit --cluster my-cluster --image=image:tag -- python job.py
+```
+
+### Ray Session - Interactive Client / Dashboard
+
+```
+$ kubectl ray cluster session my-cluster
+Connecting...
+Ray Interactive Client Session started on port 10001...
+Ray Dashboard Session started on port 8265...
+```
+
+### Ray Dashboard
+
+```
+$ kubectl ray cluster dashboard my-cluster
+Opening browser session to Ray dashboard ...
+```
+
+### Ray Logs
+
+```
+$ kubectl ray cluster logs my-cluster --out-dir ./log-dir
+Downloading Ray logs to ./log-dir ...
+```
+
+## Implementation Details
+
+The kubectl plugin will be developed in the cli directory, replacing the current KubeRay CLI. While the kubectl plugin will overlap with the existing KubeRay CLI in some ways (especially in managing KubeRay cluster), it will go further by enhancing day-to-day operations with Ray clusters. These enhancements include authenticating to clusters, establishing local sessions, submitting jobs, and scaling clusters. In addition, the kubectl plugin will not depend on the KubeRay API server, making it viable for a larger audience of KubeRay users.
```

Comment:
I am OK with deprecating the existing KubeRay CLI. I have never heard of any user using it.